book/tutorials/photon_classifier.py

```diff
@@ -0,0 +1,261 @@
+# ---
+# jupyter:
+#   jupytext:
+#     formats: py:percent
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.16.2
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Machine Learning with ICESat-2 data
+#
+# A machine learning pipeline from point clouds to photon classifications.
+#
+# Reimplementation of
+# https://github.com/YoungHyunKoo/IS2_ML/blob/main/01_find_overlapped_data.ipynb
+
+# %% [markdown]
+# ```{admonition} Learning Objectives
+# By the end of this tutorial, you should be able to:
+# - Convert ICESat-2 point cloud data into an analysis/ML-ready format
+# - Recognize the different levels of complexity of ML approaches and the
+#   benefits/challenges of each
+# - Learn the potential of using ML for ICESat-2 photon classification
+# ```
+
+# %% [markdown]
+# ## Part 0: Setup
+
+# %%
+import geopandas as gpd
+import h5py
+import icepyx as ipx
+import s3fs
+import torch
+
+# %% [markdown]
+# ## Part 1: Convert ICESat-2 data into ML-ready format
+#
+# Steps:
+# - Access using icepyx
+# - Filter to only strong beams
+# - Subset to 6 data variables only
+#
+# TODO: copy Table 1 from Koo et al., 2023 paper
+
+# %%
+# Set up spatiotemporal query for ATL07 sea ice product
+region_ross = ipx.Query(
+    product="ATL07",
+    spatial_extent=[-180, -78, -140, -70],
+    date_range=["2018-09-15", "2019-03-31"],
+    version="006",
+)
+
+# %%
+region_ross.visualize_spatial_extent()
+
+# %%
+s3links = region_ross.avail_granules(cloud=True)
+# s3links
+
+# %%
+# Authenticate using NASA EarthData login; enter your user id and password when prompted
+credentials = region_ross.s3login_credentials
+credentials
```

Comment:
Hmm, do we have some secrets in this repo for Earthdata login credentials?

```python-traceback
---------------------------------------------------------------------------
AuthenticationError                       Traceback (most recent call last)
Cell In[5], line 2
      1 # Authenticate using NASA EarthData login; enter your user id and password when prompted
----> 2 credentials = region_ross.s3login_credentials
      3 credentials

File ~/micromamba/envs/hackweek/lib/python3.11/site-packages/icepyx/core/auth.py:112, in EarthdataAuthMixin.s3login_credentials(self)
    109 # Only generate s3login_credentials the first time credentials are accessed, or if an hour
    110 # has passed since the last login
    111 if self._s3login_credentials is None:
--> 112     set_s3_creds()
    113 elif (datetime.datetime.now() - self._s3_initial_ts) >= datetime.timedelta(
    114     hours=1
    115 ):
    116     set_s3_creds()

File ~/micromamba/envs/hackweek/lib/python3.11/site-packages/icepyx/core/auth.py:106, in EarthdataAuthMixin.s3login_credentials.<locals>.set_s3_creds()
    104 def set_s3_creds():
    105     """Store s3login creds from `auth`and reset the last updated timestamp"""
--> 106     self._s3login_credentials = self.auth.get_s3_credentials(daac="NSIDC")
    107     self._s3_initial_ts = datetime.datetime.now()

File ~/micromamba/envs/hackweek/lib/python3.11/site-packages/icepyx/core/auth.py:76, in EarthdataAuthMixin.auth(self)
     74 # check for a valid auth response
     75 if auth.authenticated is False:
---> 76     raise AuthenticationError(
     77         "Earthdata authentication failed. Check output for error message"
     78     )
     79 else:
     80     self._auth = auth

AuthenticationError: Earthdata authentication failed. Check output for error message
```

I remember asking two years ago and there was a solution at https://github.com/ICESAT-2HackWeek/website2022/pull/40#issuecomment-1072890483. Will need to see what the new syntax is...

Reply:
Refactored from `icepyx` to `earthaccess` in d616d77b11d745add2bde0a561c1de45b0e2cdc1. Still getting an `AuthenticationError` though, let me open another thread.